x/evm/state/balance.go

```diff
@@ -0,0 +1,103 @@
+package state
+
+import (
+	"fmt"
+	"math"
+	"math/big"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/sei-protocol/sei-chain/x/evm/types"
+)
+
+func (s *StateDBImpl) CreateAccount(common.Address) {
+	// noop
+	// EVM account creation is handled in ante handlers.
+	// State initialization is handled in Get/SetState
+}
+
+func (s *StateDBImpl) SubBalance(evmAddr common.Address, amt *big.Int) {
+	if amt.Sign() == 0 {
+		return
+	}
+	if amt.Sign() < 0 {
+		s.AddBalance(evmAddr, new(big.Int).Neg(amt))
+		return
+	}
+	defer func() {
+		s.deficit = new(big.Int).Sub(s.deficit, amt)
+	}()
+
+	if seiAddr, ok := s.k.GetSeiAddress(s.ctx, evmAddr); ok {
+		// debit seiAddr's bank balance and credit EVM module account
+		coins := sdk.NewCoins(sdk.NewCoin(s.k.GetBaseDenom(s.ctx), sdk.NewIntFromBigInt(amt)))
+		s.err = s.k.BankKeeper().SendCoinsFromAccountToModule(s.ctx, seiAddr, types.ModuleName, coins)
+		return
+	}
+
+	balance := s.k.GetBalance(s.ctx, evmAddr)
+	if amt.Uint64() > balance {
+		s.err = fmt.Errorf("insufficient balance of %d in %s for a %s subtraction", balance, evmAddr, amt)
+		return
+	}
+
+	s.k.SetOrDeleteBalance(s.ctx, evmAddr, balance-amt.Uint64())
+}
+
+func (s *StateDBImpl) AddBalance(evmAddr common.Address, amt *big.Int) {
+	if amt.Sign() == 0 {
+		return
+	}
+	if amt.Sign() < 0 {
+		s.SubBalance(evmAddr, new(big.Int).Neg(amt))
+		return
+	}
+	defer func() {
+		s.deficit = new(big.Int).Add(s.deficit, amt)
+	}()
+
+	if seiAddr, ok := s.k.GetSeiAddress(s.ctx, evmAddr); ok {
+		// credit seiAddr's bank balance and debit EVM module account, mint if needed
```

Comment:
can you walk me through this? are we minting sei coins if the evm module account doesn't have enough balance?

Reply:
yeah this would only happen if during a money movement in EVM, an `AddBalance` is called before the corresponding `SubBalance`. Since evm module account hasn't received the funds from `SubBalance` yet, it needs to temporarily mint some to cover the needs in `AddBalance`. All minted tokens will be burned in `CheckBalance` which is called once at the end of each EVM transaction